File: packages/api-types/src/ws/dto/agent-chat.ts
export type AgentChatOrigin = "quick" | "normal";

export type AgentChatPrefs = {
  last_registry_id?: string | null;
};

export type AgentChatPrefsSetRequest = {
  last_registry_id?: string | null;
};

export type AgentChatCreateRequest = {
  workspace_id?: string | null;
  project_id?: string | null;
  space_id?: string | null;
  cwd?: string | null;
  provider_id: string;
  model?: string | null;
  thinking?: string | null;
  mode?: string | null;
  title?: string | null;
  origin?: AgentChatOrigin | null;
};

export type AgentChatListRequest = {
  workspace_id?: string | null;
  project_id?: string | null;
  cwd?: string | null;
  cursor?: string | null;
  limit?: number | null;
  all?: boolean | null;
  origin?: AgentChatOrigin | null;
};

export type AgentChatIdRequest = {
  chat_id: string;
};

export type AgentChatMessagesRequest = {
  chat_id: string;
  limit?: number | null;
};

export type AgentChatRenameRequest = {
  chat_id: string;
  title: string;
};

export type AgentChatConfigureRequest = {
  chat_id: string;
  provider_id?: string | null;
  model?: string | null;
  thinking?: string | null;
  mode?: string | null;
  permission_mode?: string | null;
};

export type AgentChatSubscribeRequest = {
  chat_id: string;
  after_sequence?: number | null;
};

export type AgentChatSendRequest = {
  chat_id: string;
  text: string;
  attachment_paths?: string[] | null;
};

export type AgentChatSteerRequest = {
  chat_id: string;
  expected_turn_id: string;
  text: string;
};

export type AgentChatQueueAddRequest = {
  chat_id: string;
  text: string;
  attachment_paths?: string[] | null;
};

export type AgentChatQueueUpdateRequest = {
  chat_id: string;
  item_id: string;
  text?: string | null;
  status?: string | null;
};

export type AgentChatQueueReorderRequest = {
  chat_id: string;
  item_ids: string[];
};

export type AgentChatQueueDeleteRequest = {
  chat_id: string;
  item_id: string;
};

export type AgentChatPermissionRespondRequest = {
  chat_id: string;
  request_id: string;
  option_id?: string | null;
  allowed?: boolean | null;
};

export type AgentChatSessionOpRespondRequest = {
  chat_id: string;
  request_id: string;
  option_id: string;
};

export type AgentModelCatalogGetRequest = {
  agent_id: string;
  refresh?: boolean | null;
};

export type Capability = "supported" | "unsupported";

export type AgentThinkingSupport =
  | { type: "none" }
  | { type: "enum"; arg?: string | null; options: string[] }
  | { type: "manual"; arg: string; placeholder?: string | null }
  | { type: "encoded_in_model" }
  | { type: "flag_only"; arg: string };

export type AgentIdentity = {
  id: string;
  name: string;
  version?: string | null;
};

export type AgentCapabilities = {
  steer: Capability;
  resume: Capability;
  permission: Capability;
  configure: Capability;
  fork: Capability;
  rewind: Capability;
};

export type AgentOptionSupport = {
  models: Capability;
  thinking: Capability;
  modes: Capability;
  permission_modes: Capability;
};

export type AgentDescriptor = {
  identity: AgentIdentity;
  capabilities: AgentCapabilities;
  support: AgentOptionSupport;
  supported_options: {
    models: Array<{
      id: string;
      label: string;
      group?: string | null;
      is_default?: boolean;
      thinking?: AgentThinkingSupport | null;
    }>;
    thinking?: AgentThinkingSupport;
    modes?: Array<{ id: string; label: string; is_default?: boolean }>;
    permission_modes?: Array<{ id: string; label: string; is_default?: boolean }>;
  };
  current_config: {
    model?: string | null;
    thinking?: string | null;
    mode?: string | null;
    permission_mode?: string | null;
  };
};

export type AgentChatMeta = {
  id: string;
  created_at: string;
  updated_at: string;
  deleted: boolean;
  title: string | null;
  cwd: string;
  workspace_id: string | null;
  project_id: string | null;
  space_id?: string | null;
  origin?: AgentChatOrigin | null;
  provider_id: string;
  last_message_at: string | null;
  last_event_seq: number;
  persistence_handle: string | null;
  runtime_status: string;
  available_commands?: Array<{
    name: string;
    description: string;
    hint?: string | null;
  }>;
  session_usage?: AgentSessionUsage | null;
  descriptor: AgentDescriptor;
  parent_chat_id?: string | null;
  rewind_view?: { until_turn_id: string } | null;
};

export type AgentSessionUsage = {
  used?: number | null;
  size?: number | null;
  cost?: {
    amount?: number | null;
    currency?: string | null;
  } | null;
};

export type AgentTurnUsage = {
  total_tokens?: number | null;
  input_tokens?: number | null;
  output_tokens?: number | null;
  thought_tokens?: number | null;
  cached_read_tokens?: number | null;
  cached_write_tokens?: number | null;
};

export type AgentChatIndexEntry = {
  id: string;
  title: string | null;
  cwd: string;
  workspace_id: string | null;
  project_id: string | null;
  space_id?: string | null;
  origin?: AgentChatOrigin | null;
  provider_id: string;
  updated_at: string;
  last_message_at: string | null;
  deleted: boolean;
};

export type AgentChatListResponse = {
  items: AgentChatIndexEntry[];
};

export type AgentToolKind =
  | "read"
  | "edit"
  | "delete"
  | "move"
  | "search"
  | "web_search"
  | "execute"
  | "fetch"
  | "skill"
  | "subagent"
  | "other";

export type AgentToolStatus = "pending" | "running" | "completed" | "failed";

/** Internally tagged `{ type: kind, … }`. `other.value` is vendor JSON as-is. */
export type AgentToolParams =
  | { type: "read"; path: string; offset?: number | null; limit?: number | null }
  | { type: "edit"; path: string }
  | { type: "delete"; path: string }
  | { type: "move"; from: string; to: string }
  | { type: "search"; query: string; path?: string | null; glob?: string | null }
  | { type: "web_search"; query: string }
  | {
      type: "execute";
      command: string;
      cwd?: string | null;
      background: boolean;
      task_id?: string | null;
    }
  | { type: "fetch"; url: string }
  | { type: "skill"; skill: string }
  | { type: "subagent"; description: string; agent_type?: string | null }
  | { type: "other"; value: unknown };

export type AgentToolResult =
  | { type: "text"; text: string }
  | { type: "file_content"; path: string; text: string }
  | { type: "diff_stats"; path: string; additions: number; deletions: number }
  | { type: "execute"; output: string; exit_code?: number | null }
  | {
      type: "web_search";
      query: string;
      links: Array<{ url: string; title: string; snippet?: string | null }>;
    }
  | {
      type: "search_hits";
      query: string;
      hits: Array<{
        path: string;
        line?: number | null;
        snippet?: string | null;
      }>;
    }
  | {
      type: "web_fetch";
      url: string;
      title?: string | null;
      markdown?: string | null;
      text?: string | null;
    }
  | { type: "other"; value: unknown }
  | { type: "error"; message: string }
  | { type: "empty" };

export type AgentTool = {
  tool_call_id: string;
  name: string;
  title?: string | null;
  kind: AgentToolKind;
  status: AgentToolStatus;
  params: AgentToolParams;
  result?: AgentToolResult | null;
};

export type SessionLifecycleAction = "create" | "resume";
export type SessionLifecycleStatus = "running" | "completed" | "failed";
export type SessionHintTone = "info" | "warning" | "error";

export type SessionConfigValueChange = {
  from?: string | null;
  to: string;
};

export type AgentPart =
  | { type: "text"; text: string }
  | { type: "thinking"; text: string; tool_call_id?: string; duration_ms?: number | null }
  | ({ type: "tool_call" } & AgentTool)
  | { type: "plan"; plan: unknown }
  | { type: "attachment"; path: string; name?: string | null }
  | { type: "error"; message: string }
  | {
      type: "session_lifecycle";
      action: SessionLifecycleAction;
      status: SessionLifecycleStatus;
      duration_ms?: number | null;
      error?: string | null;
    }
  | {
      type: "session_config_change";
      model?: SessionConfigValueChange | null;
      mode?: SessionConfigValueChange | null;
    }
  | {
      type: "session_hint";
      tone: SessionHintTone;
      kind: string;
    };

export type AgentMessage = {
  id: string;
  role: "user" | "assistant" | string;
  kind?: "normal" | "steer" | string;
  parts: AgentPart[];
  created_at?: string;
  streaming?: boolean;
  worked_ms?: number | null;
  thinking_ms?: number | null;
  completed_at?: string | null;
  usage?: AgentTurnUsage | null;
};

export type AgentQueueItem = {
  id: string;
  seq: number;
  status: string;
  prompt: string;
  display_prompt?: string | null;
  attachments?: string[];
};

export type AgentSessionOpKind = "fork" | "rewind";

export type AgentSessionOpOutcome = "applied" | "canceled" | "failed";

export type AgentSessionOpRequest = {
  request_id: string;
  kind: AgentSessionOpKind;
  title: string;
  options: Array<{ option_id: string; name: string; kind?: string }>;
};

export type AgentChatSnapshot = {
  meta: AgentChatMeta;
  messages: AgentMessage[];
  queue: AgentQueueItem[];
  pending_permission?: {
    request_id: string;
    tool: string;
    description: string;
    content_markdown?: string | null;
    options?: Array<{ option_id: string; name: string; kind?: string }>;
    status: string;
  } | null;
  pending_session_op?: AgentSessionOpRequest | null;
  running_turn_id?: string | null;
  running_turn_started_at?: string | null;
};

export type AgentChatTurnIdResponse = {
  turn_id: string;
};

export type AgentChatSubscribeResponse = {
  last_event_seq: number;
};

export type AgentChatPayload =
  | { type: "turn_started"; turn_id: string; created_at?: string | null }
  | {
      type: "user_message";
      turn_id: string;
      message_id: string;
      kind?: string;
      text: string;
      attachments?: string[];
      created_at?: string;
    }
  | { type: "assistant_message_delta"; message_id: string; delta: string; turn_id?: string }
  | { type: "assistant_message_completed"; message_id: string }
  | { type: "thinking_delta"; message_id: string; delta: string; turn_id?: string }
  | { type: "thinking_completed"; message_id: string; thinking_ms?: number | null }
  | {
      type: "tool_call_started" | "tool_call_updated" | "tool_call_completed";
      tool_call: AgentTool;
    }
  | {
      type: "tool_call_failed";
      tool_call: AgentTool;
      error?: string | null;
    }
  | { type: "plan_updated"; plan?: unknown; turn_id?: string }
  | {
      type: "permission_requested";
      request: {
        request_id: string;
        tool?: string;
        description?: string;
        content_markdown?: string;
        options?: Array<{ option_id: string; name: string; kind?: string }>;
      };
    }
  | { type: "permission_resolved"; request_id: string; option_id: string }
  | { type: "session_op_requested"; request: AgentSessionOpRequest }
  | {
      type: "session_op_resolved";
      request_id: string;
      option_id: string;
      outcome: AgentSessionOpOutcome;
      error?: string | null;
    }
  | { type: "session_forked"; parent_chat_id: string; chat_id: string }
  | { type: "rewind_view_updated"; until_turn_id: string | null }
  | {
      type: "turn_completed";
      turn_id: string;
      status?: string;
      worked_ms?: number | null;
      thinking_ms?: number | null;
      completed_at?: string | null;
      usage?: AgentTurnUsage | null;
      error?: string | null;
    }
  | {
      type: "usage_updated";
      session?: AgentSessionUsage | null;
      turn?: AgentTurnUsage | null;
    }
  | {
      type: "queue_updated";
      items: Array<{
        id: string;
        seq: number;
        status: string;
        prompt: string;
        display_prompt?: string | null;
        attachments?: string[];
      }>;
    }
  | { type: "runtime_status"; status?: string; persistence_handle?: string | null }
  | { type: "title_updated"; title?: string | null }
  | {
      type: "available_commands_updated";
      commands?: Array<{ name?: string; description?: string; hint?: string | null }>;
    }
  | { type: "config_updated"; descriptor: AgentDescriptor }
  | { type: "unknown"; event_type: string; payload: unknown }
  | {
      type: "session_lifecycle";
      turn_id: string;
      message_id: string;
      action: SessionLifecycleAction;
      status: SessionLifecycleStatus;
      duration_ms?: number | null;
      error?: string | null;
    }
  | {
      type: "session_config_change";
      turn_id: string;
      message_id: string;
      model?: SessionConfigValueChange | null;
      mode?: SessionConfigValueChange | null;
    }
  | {
      type: "session_hint";
      turn_id: string;
      message_id: string;
      tone: SessionHintTone;
      kind: string;
    };

/** Host event log item. Same tagged union the server emits on `agent_chat_event`. */
export type AgentEvent = AgentChatPayload;

export type AgentChatEvent = {
  chat_id: string;
  event_id: string;
  sequence: number;
  turn_id?: string | null;
  payload: AgentEvent;
};

export type AgentModelCatalog = {
  agent_id: string;
  status: "ok" | "unsupported" | "auth_required" | "error" | "probing";
  models: Array<{
    id: string;
    label: string;
    group?: string | null;
    is_default?: boolean;
    thinking?: AgentThinkingSupport | null;
  }>;
  modes: Array<{ id: string; label: string; is_default?: boolean }>;
  permission_modes?: Array<{ id: string; label: string; is_default?: boolean }>;
  thinking: AgentThinkingSupport;
  strategies_used: string[];
  fetched_at: string;
  source: "cache" | "live";
  message: string | null;
  commands?: Array<{
    name: string;
    description: string;
    hint?: string | null;
  }>;
};

export type AgentModelCatalogUpdated = {
  agent_id: string;
  catalog: AgentModelCatalog;
};

File: apps/api/src/api/ws/message/agent_chat.rs
use agent::{AgentAvailableCommand, AgentDescriptor};
use chrono::{DateTime, Utc};
use core_service::service::agent_chat::SessionUsage;
use core_service::{AgentChatMeta, AgentChatOrigin, AgentChatSnapshot, RuntimeStatus};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatCreateRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub space_id: Option<String>,
    #[serde(default)]
    pub cwd: Option<String>,
    pub provider_id: String,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub thinking: Option<String>,
    #[serde(default)]
    pub mode: Option<String>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub origin: Option<AgentChatOrigin>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatListRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub cwd: Option<String>,
    #[serde(default)]
    pub cursor: Option<String>,
    #[serde(default)]
    pub limit: Option<u32>,
    /// When true, unscoped list returns every chat instead of scratch-only.
    #[serde(default)]
    pub all: bool,
    #[serde(default)]
    pub origin: Option<AgentChatOrigin>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatIdRequest {
    pub chat_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatMessagesRequest {
    pub chat_id: String,
    #[serde(default)]
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatRenameRequest {
    pub chat_id: String,
    pub title: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatSubscribeRequest {
    pub chat_id: String,
    #[serde(default)]
    pub after_sequence: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatSendRequest {
    pub chat_id: String,
    pub text: String,
    #[serde(default)]
    pub attachment_paths: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatSteerRequest {
    pub chat_id: String,
    pub expected_turn_id: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatQueueAddRequest {
    pub chat_id: String,
    pub text: String,
    #[serde(default)]
    pub attachment_paths: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatQueueUpdateRequest {
    pub chat_id: String,
    pub item_id: String,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatQueueReorderRequest {
    pub chat_id: String,
    pub item_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatQueueDeleteRequest {
    pub chat_id: String,
    pub item_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatPermissionRespondRequest {
    pub chat_id: String,
    pub request_id: String,
    #[serde(default)]
    pub option_id: Option<String>,
    #[serde(default)]
    pub allowed: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatSessionOpRespondRequest {
    pub chat_id: String,
    pub request_id: String,
    pub option_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentModelCatalogGetRequest {
    pub agent_id: String,
    #[serde(default)]
    pub refresh: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatOk {
    pub ok: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentChatConfigureRequest {
    pub chat_id: String,
    #[serde(default)]
    pub provider_id: Option<String>,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub thinking: Option<String>,
    #[serde(default)]
    pub mode: Option<String>,
    #[serde(default)]
    pub permission_mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AgentChatPrefsSetRequest {
    #[serde(default)]
    pub last_registry_id: Option<String>,
}

/// WS JSON for `AgentChatMeta`. Persist-only `applied_*` stay off the wire.
#[derive(Debug, Clone, Serialize)]
pub struct AgentChatMetaWire {
    pub id: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub deleted: bool,
    pub title: Option<String>,
    pub cwd: String,
    pub workspace_id: Option<String>,
    pub project_id: Option<String>,
    pub space_id: Option<String>,
    pub origin: AgentChatOrigin,
    pub provider_id: String,
    pub last_message_at: Option<DateTime<Utc>>,
    pub last_event_seq: u64,
    pub persistence_handle: Option<String>,
    pub runtime_status: RuntimeStatus,
    pub available_commands: Vec<AgentAvailableCommand>,
    pub session_usage: Option<SessionUsage>,
    pub descriptor: AgentDescriptor,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_chat_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rewind_view: Option<AgentChatRewindViewWire>,
}

#[derive(Debug, Clone, Serialize)]
pub struct AgentChatRewindViewWire {
    pub until_turn_id: String,
}

impl From<&AgentChatMeta> for AgentChatMetaWire {
    fn from(meta: &AgentChatMeta) -> Self {
        Self {
            id: meta.id.clone(),
            created_at: meta.created_at,
            updated_at: meta.updated_at,
            deleted: meta.deleted,
            title: meta.title.clone(),
            cwd: meta.cwd.clone(),
            workspace_id: meta.workspace_id.clone(),
            project_id: meta.project_id.clone(),
            space_id: meta.space_id.clone(),
            origin: meta.origin,
            provider_id: meta.provider_id.clone(),
            last_message_at: meta.last_message_at,
            last_event_seq: meta.last_event_seq,
            persistence_handle: meta.persistence_handle.clone(),
            runtime_status: meta.runtime_status,
            available_commands: meta.available_commands.clone(),
            session_usage: meta.session_usage.clone(),
            descriptor: meta.descriptor.clone(),
            parent_chat_id: meta.parent_chat_id.clone(),
            rewind_view: meta
                .rewind_view
                .as_ref()
                .map(|view| AgentChatRewindViewWire {
                    until_turn_id: view.until_turn_id.clone(),
                }),
        }
    }
}

pub fn agent_chat_meta_json(meta: &AgentChatMeta) -> Result<Value, serde_json::Error> {
    serde_json::to_value(AgentChatMetaWire::from(meta))
}

pub fn agent_chat_snapshot_json(snapshot: &AgentChatSnapshot) -> Result<Value, serde_json::Error> {
    let mut value = serde_json::to_value(snapshot)?;
    value["meta"] = serde_json::to_value(AgentChatMetaWire::from(&snapshot.meta))?;
    Ok(value)
}

#[cfg(test)]
mod wire_tests {
    use super::*;
    use agent::{AgentCurrentConfig, AgentIdentity, AgentSupportedOptions};

    fn meta_with_applied() -> AgentChatMeta {
        AgentChatMeta {
            id: "chat-1".into(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
            deleted: false,
            title: Some("hello".into()),
            cwd: "/tmp".into(),
            workspace_id: None,
            project_id: None,
            space_id: None,
            origin: AgentChatOrigin::Normal,
            provider_id: "claude".into(),
            last_message_at: None,
            last_event_seq: 3,
            persistence_handle: None,
            runtime_status: RuntimeStatus::Detached,
            applied_model: Some("opus".into()),
            applied_thinking: Some("high".into()),
            applied_mode: Some("plan".into()),
            applied_permission_mode: None,
            available_commands: Vec::new(),
            session_usage: None,
            descriptor: AgentDescriptor {
                identity: AgentIdentity {
                    id: "claude".into(),
                    name: "Claude Code".into(),
                    version: None,
                },
                capabilities: agent::capabilities_for_provider("claude"),
                support: agent::option_support_for_provider("claude"),
                supported_options: AgentSupportedOptions::default(),
                current_config: AgentCurrentConfig {
                    model: Some("opus".into()),
                    thinking: Some("high".into()),
                    mode: None,
                    ..AgentCurrentConfig::default()
                },
            },
            parent_chat_id: None,
            rewind_view: None,
            pending_session_op: None,
        }
    }

    #[test]
    fn meta_wire_omits_persist_only_and_legacy_keys() {
        let meta = meta_with_applied();
        let value = agent_chat_meta_json(&meta).expect("serialize");
        let object = value.as_object().expect("object");
        assert!(object.get("applied_model").is_none());
        assert!(object.get("applied_thinking").is_none());
        assert!(object.get("applied_mode").is_none());
        assert!(object.get("supports_steer").is_none());
        assert!(object.get("selected_model").is_none());
        assert!(object.get("selected_thinking").is_none());
        assert!(object.get("selected_mode").is_none());
        assert!(object.get("session_config_options").is_none());
        assert_eq!(
            object
                .get("descriptor")
                .and_then(|item| item.get("identity"))
                .and_then(|item| item.get("id"))
                .and_then(|item| item.as_str()),
            Some("claude")
        );
        assert_eq!(
            object.get("provider_id").and_then(|item| item.as_str()),
            Some("claude")
        );
        assert_eq!(
            object
                .get("descriptor")
                .and_then(|item| item.get("current_config"))
                .and_then(|item| item.get("model"))
                .and_then(|item| item.as_str()),
            Some("opus")
        );
    }

    #[test]
    fn persist_meta_still_has_applied_fields() {
        let value = serde_json::to_value(meta_with_applied()).expect("persist");
        let object = value.as_object().expect("object");
        assert_eq!(
            object.get("applied_model").and_then(|item| item.as_str()),
            Some("opus")
        );
        assert!(object.get("descriptor").is_some());
    }

    #[test]
    fn snapshot_wire_strips_applied_from_nested_meta() {
        let snapshot = AgentChatSnapshot {
            meta: meta_with_applied(),
            messages: Vec::new(),
            queue: Vec::new(),
            pending_permission: None,
            running_turn_id: None,
            running_turn_started_at: None,
        };
        let value = agent_chat_snapshot_json(&snapshot).expect("serialize");
        let meta = value
            .get("meta")
            .and_then(|item| item.as_object())
            .expect("meta");
        assert!(meta.get("applied_model").is_none());
        assert!(meta.get("applied_thinking").is_none());
        assert!(meta.get("applied_mode").is_none());
        assert!(meta.get("descriptor").is_some());
        assert!(value.get("messages").is_some());
    }
}
